### ISSUE TITLE
Cusolver handle may decrease MAGMA performance on GPU

### ISSUE THREAD

**Issue by xwang233** [pytorch/pytorch#55122]
## 🐛 Bug

There are some weird interactions between cusolver handle and MAGMA performance. 🤔 

I have uploaded related files to https://github.com/xwang233/code-snippet/tree/master/magma-cusolver-handle.

## What happened?

Currently in pytorch, matrix inverse of batched matrix with `batch size <= 2` is dispatched to cusolver; matrix inverse of batched matrix with `batch size > 2` is dispatched to magma. This heuristic is added to pick the better performance between the two backends for different matrix sizes.

Cusolver handles are created when a cusolver call is needed. After the cusolver call, the handle is released into a handle pool instead of being destroyed. This is because cusolver handle creation and destroy take very long time, so reuse them gives better performance in most cases. Cublas, cusparse, cudnn handles in pytorch follow the same "handle pool" implementation.

Ideally, the two backends shouldn't affect performance of each other. However, it turns out that if someone runs matrix inverse in the following way:

```
magma inverse (A), then cusolver inverse (B), then magma inverse (C)
```

They will see that magma inverse performance in (C) is worse than that of (A), even for exactly the same matrix and exactly the same python statements.

Run this script (https://github.com/xwang233/code-snippet/blob/master/magma-cusolver-handle/a.py) to profile `torch.inverse`, and see the results yourself! You may need pytorch version >= 1.7.0 to see it. Cusolver inverse was initially added in #42403. 

Example output (per iteration time in ms, smaller is faster)

```
1.9.0a0+gitfbaad8c
time_magma_1 =  0.184
time_magma_2 =  0.411
```

The MAGMA inverse call after cusolver is regressed for about 0.23 ms per iteration.

## Why this happened?

I've checked the performance with Nvidia Nsight Systems, and here are the two different MAGMA inverse reports:

Magma inverse call in (A):

![first magma](https://raw.githubusercontent.com/xwang233/code-snippet/master/magma-cusolver-handle/first-magma.png)

Magma inverse call in (C):

![second magma](https://raw.githubusercontent.com/xwang233/code-snippet/master/magma-cusolver-handle/second-magma.png)

It is found that the `cudaMalloc` and `cudaFree` in second magma (C) call takes longer time than first magma (A) call. It is also found that there are those extra OSRT library calls `ioctl`. Sum of the 4 `ioctl` blocks gives a time of ~210 us, which is in the same order of magnitude of the regressed performance. (However, we don't know if they have causal relations.)

## Where does this `ioctl` come from?

See stack trace 

![ioctl stacktrace](https://raw.githubusercontent.com/xwang233/code-snippet/master/magma-cusolver-handle/second-magma-ioctl-1-trace.png)

It's probably from a constructor call here

https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/aten/src/ATen/native/cuda/MiscUtils.h#L35-L41

FYI, the third argument here is a cublas handle. So magma queue creation takes cuda stream, cublas handle, cusparse handle, and has nothing to do with cusolver handle.

## Why is cusolver related to this MAGMA queue creation? 

Good question, and a tough one. To save you some time on debug, I wrote a short C++ repro program

https://github.com/xwang233/code-snippet/blob/dcc269de7d1fba131ffd747ba15b1feb5f09a527/magma-cusolver-handle/cpp/main.cpp#L274-L283

The program simply runs and profiles MAGMA `magma_sgetrf_batched` + `magma_sgetri_outofplace_batched` calls in three different "environments"

- normal
- a cusolver handle is created (not even used in any cusolver functions)
- the cusolver handle is destroyed 

The three environments correspond to the (A), (B), (C) cases above. Here are sample outputs

```
magma time elapsed 0.063466 ms
magma time elapsed 0.272094 ms
magma time elapsed 0.063150 ms
```

It shows that, somehow, creation of a cusolver handle may decrease MAGMA performance.

## Which pytorch operators are affected?

- inverse
- cholesky
- cholesky_solve (PR ongoing)
- cholesky_inverse (PR ongoing, it calls cholesky_solve as its implementation)

There may be other linear algebra operators, that use MAGMA backend, affected, but I haven't check benchmark of them yet. Also note that the process (B) doesn't have to be the same linear algebra call as (A) or (C). It could be any cusolver call that creates a cusolver handle.

This regression seems to be a fixed duration of ~0.2 ms on my machine, which means it's mostly negligible for large matrix operations. 

## Additional context

See also #42403 #42666 #47953

cc @ptrblck @ngimel @mruberry @heitorschueroff @IvanYashchuk

**Comment by ngimel** [comment]
Thank you @xwang233, this is very helpful! So unfortunately calls the do have `queue` arguments are that are successfully reusing (instead of creating) cublas and cusparse handles are subject to this. 
Here are 3 magma_malloc calls (that call cudaMalloc in turn) https://bitbucket.org/icl/magma/src/6e3a460f9badffb1c391b2d34bb1477ad8eb7367/interface_cuda/interface.cpp#lines-1128:1133, there's nothing special about them. Are you seeing only the third one being longer? 
Perhaps cuda team might know something about why cudaMalloc would be slower under some circumstances?
And yes, ideally we should not be calling cudaMalloc and corresponding cudaFree, but that's not possible without modifying magma source.

**Comment by xwang233** [comment]
Thanks @ngimel . I think the lines in magma source you pointed to are for `magma_queue_create_from_hip_internal`, and there is another `magma_queue_create_from_cuda_internal` above in the same file. I'm still not quite sure where the `cudaMalloc` comes from.

> Perhaps cuda team might know something about why cudaMalloc would be slower under some circumstances?

Sure, I can check with them and see if there is anything.

In my local testing, I was thinking if magma_queue can be reused. I'm just not very familiar with magma mechanism. Perhaps @IvanYashchuk knows more.

**Comment by ngimel** [comment]
@xwang233 you are right, those lines are from `magma_queue_create_from_hip` and `magma_queue_create_from_cuda_internal` doesn't even have magma_alloc calls, this is a mystery.

**Comment by xwang233** [comment]
A quick update. 

For the C++ code here https://github.com/xwang233/code-snippet/blob/dcc269de7d1fba131ffd747ba15b1feb5f09a527/magma-cusolver-handle/cpp/main.cpp#L274-L283, I've replaced the cusolver handle create/destroy to simple `cudaMalloc` and `cudaFree` calls of 10^6 byte. I can still see the same regression in MAGMA inverse (or queue creation maybe). 

This probably means that the cause of MAGMA regression is not cusolver. It's more likely any `cudaMalloc` ahead of MAGMA queue creation. I'm still trying to get a python repro as well.

Cuda team mentioned that `cudaMalloc` may call `ioctl` internally if the size allocated is too large. On the other hand, previously allocated space may push the new `cudaMalloc` beyond a boundary, where `ioctl` calls are then needed.

@ngimel my local environment is using MAGMA 2.5.3, which seems to still have the `magma_malloc`, https://bitbucket.org/icl/magma/src/40353c62e1f85e416347242d38583a3e1e25b5ea/interface_cuda/interface.cpp#lines-896:901. I haven't check if 2.5.4 would solve the issue.

**Comment by xwang233** [comment]
Well, good news and bad news. 

I built MAGMA 2.5.4 and linked that with pytorch. Using my script `a.py` in cuda 11.1, I still see the same "order of magnitude" of performance regression in matrix inverse, so I checked the nsys report.

The three `cudaMalloc` at the beginning in magma queue creation no longer exist, but another `cudaMalloc` from `magma_sgetri_outofplace_batched` is regressed with `ioctl`. 😂 

It seems like, if a MAGMA call has `magma_malloc` in it, the regression is mostly inevitable. Whether there is a smart workaround on pytorch side (C++) to fix this would be another story.

![second magma](https://raw.githubusercontent.com/xwang233/code-snippet/master/magma-cusolver-handle/second-magma_2.5.4.png)

![second magma ioctl](https://raw.githubusercontent.com/xwang233/code-snippet/master/magma-cusolver-handle/second-magma-ioctl-trace_2.5.4.png)

**Comment by ngimel** [comment]
Damn, ok, this means we transitively use magma_sgemm_batched through magma_sgetri_batched :-( I thought we don't use it at all. I'm out of good ideas :-( There are more invasive changes to magma that'd avoid that, but it means that we'd need to nontrivially patch magma every time we update the version.
